foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/helper/ConcurrencyManager.java

```diff
@@ -582,4 +757,130 @@ private static String getPropertyRecordStackOnLock() {
+
+    /** Getter for {@link #THREADS_WAITING_TO_RELEASE_DEFERRED_LOCKS} */
+    public static Set<Thread> getThreadsWaitingToReleaseDeferredLocks() {
+        return new HashSet<>(THREADS_WAITING_TO_RELEASE_DEFERRED_LOCKS);
+    }
+
+    /**
+     * The current thread has incremented the number of readers on the current cache key. It also wants to record into
+     * the read lock manager that this thread has acquired the cache key. This method should be user in all places where
+     * the cache key nunber of readers is incremented.
+     */
+    protected void addReadLockToReadLockManager() {
+        Thread currentThread = Thread.currentThread();
+        ReadLockManager readLockManager = getReadLockManagerEnsureResultIsNotNull(currentThread);
+        ConcurrencyManager concurrencyManagerCacheKey = this;
+        readLockManager.addReadLock(concurrencyManagerCacheKey);
+    }
+
+    /**
+     * The current thread is about to decrement the number of readers in cache key. The thread also wants to update the
+     * read lock manager and remove the cache key that has previously been aquired from there.
+     */
+    protected void removeReadLockFromReadLockManager() {
+        Thread currentThread = Thread.currentThread();
+        ReadLockManager readLockManager = getReadLockManager(currentThread);
+        if (readLockManager != null) {
+            ConcurrencyManager concurrencyManagerCacheKey = this;
+            readLockManager.removeReadLock(concurrencyManagerCacheKey);
+            removeReadLockManagerIfEmpty(currentThread);
+        } else {
+            // We have a problem we do not want ever see a decrement on the number of readers if we
+            // are not tracing one or more precessro add read lock keys.
+            // so we will put the error message into a fresh new read lock manager
+            final int currentNumberOfReaders = this.numberOfReaders.get();
+            final int decrementedNumberOfReaders = currentNumberOfReaders - 1;
+            String errorMessage = ConcurrencyUtil.SINGLETON.readLockManagerProblem01CreateLogErrorMessageToIndicateThatCurrentThreadHasNullReadLockManagerWhileDecrementingNumberOfReaders(currentNumberOfReaders, decrementedNumberOfReaders, this);
+            readLockManager = getReadLockManagerEnsureResultIsNotNull(currentThread);
+            readLockManager.addRemoveReadLockProblemsDetected(errorMessage);
+        }
+    }
+
+    /**
+     * Same as {@link #getReadLockManager(Thread)} but in this case a not null result is ensured
+     *
+     * @param thread
+     *            the thread wanting its read lock manager
+     * @return the read lock manager for the current thread.
+     */
+    protected static ReadLockManager getReadLockManagerEnsureResultIsNotNull(Thread thread) {
+        Map<Thread, ReadLockManager> readLockManagers = getReadLockManagers();
+        if (!readLockManagers.containsKey(thread)) {
+            readLockManagers.putIfAbsent(thread, new ReadLockManager());
```

Comment:
would it make sense to directly return newly created `ReadLockManager` from here instead of doing additional `get` later?